Changes in module 'nf-core/biobambam/bamsormadup'
--- modules/nf-core/biobambam/bamsormadup/main.nf
+++ modules/nf-core/biobambam/bamsormadup/main.nf
@@ -29,6 +29,8 @@
     bamcat \\
         I=${input_string} \\
         level=0 \\
+    | bamcollate2 \\
+        level=0 \\
     | bamsormadup \\
         $args \\
         M=${prefix}.metrics.txt \\
@@ -39,6 +41,7 @@
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         bamcat: \$(echo \$(bamcat --version 2>&1) | sed 's/^This is biobambam2 version //; s/..biobambam2 is .*\$//' )
+        bamcollate2: \$(echo \$(bamcat --version 2>&1) | sed 's/^This is biobambam2 version //; s/..biobambam2 is .*\$//' )
         bamsormadup: \$(echo \$(bamsormadup --version 2>&1) | sed 's/^This is biobambam2 version //; s/..biobambam2 is .*\$//' )
     END_VERSIONS
     """

************************************************************
