Changes in module 'nf-core/samtools/stats'
--- modules/nf-core/samtools/stats/main.nf
+++ modules/nf-core/samtools/stats/main.nf
@@ -8,8 +8,7 @@
         'biocontainers/samtools:1.19.2--h50ea8bc_0' }"
 
     input:
-    tuple val(meta), path(input), path(input_index)
-    tuple val(meta2), path(fasta)
+    tuple val(meta), path(input), path(input_index), path(fasta)
 
     output:
     tuple val(meta), path("*.stats"), emit: stats

************************************************************
