Changes in module 'nf-core/picard/collectwgsmetrics'
--- modules/nf-core/picard/collectwgsmetrics/main.nf
+++ modules/nf-core/picard/collectwgsmetrics/main.nf
@@ -8,9 +8,7 @@
         'biocontainers/picard:3.1.1--hdfd78af_0' }"
 
     input:
-    tuple val(meta), path(bam), path(bai)
-    tuple val(meta2), path(fasta)
-    tuple val(meta3), path(fai)
+    tuple val(meta), path(bam), path(bai) ,path(fasta) ,path(fai)
     path  intervallist
 
     output:

************************************************************
