Changes in module 'nf-core/picard/collectmultiplemetrics'
--- modules/nf-core/picard/collectmultiplemetrics/main.nf
+++ modules/nf-core/picard/collectmultiplemetrics/main.nf
@@ -8,9 +8,7 @@
         'biocontainers/picard:3.2.0--hdfd78af_0' }"
 
     input:
-    tuple val(meta) , path(bam), path(bai)
-    tuple val(meta2), path(fasta)
-    tuple val(meta3), path(fai)
+    tuple val(meta) , path(bam), path(bai) ,path(fasta) ,path(fai)
 
     output:
     tuple val(meta), path("*_metrics"), emit: metrics

************************************************************
