Changes in module 'nf-core/dragmap/align'
--- modules/nf-core/dragmap/align/main.nf
+++ modules/nf-core/dragmap/align/main.nf
@@ -8,8 +8,7 @@
         'biocontainers/mulled-v2-580d344d9d4a496cd403932da8765f9e0187774d:7eed251370ac7f3537c3d9472cdb2f9f5d8da1c5-0' }"
 
     input:
-    tuple val(meta) , path(reads)
-    tuple val(meta2), path(hashmap)
+    tuple val(meta) , path(reads), path(hashmap)
     val   sort_bam
 
     output:

************************************************************
