Changes in module 'nf-core/picard/collecthsmetrics'
--- modules/nf-core/picard/collecthsmetrics/main.nf
+++ modules/nf-core/picard/collecthsmetrics/main.nf
@@ -11,6 +11,8 @@
     tuple val(meta), path(bam), path(bai), path(bait_intervals), path(target_intervals)
     tuple val(meta2), path(fasta)
     tuple val(meta3), path(fai)
+    tuple val(meta4), path(dict)
+
 
     output:
     tuple val(meta), path("*_metrics")  , emit: metrics
@@ -31,13 +33,47 @@
         avail_mem = (task.memory.mega*0.8).intValue()
     }
     """
+
+    if [[ "$bait_intervals" == *.bed || "$bait_intervals" == *.bed.gz ]]; then
+    bait_bed=$bait_intervals
+    bait_intervals=\${bait_bed%.bed*}.interval_list
+
+    picard \\
+        -Xmx${avail_mem}M \\
+        BedToIntervalList \\
+        --INPUT \$bait_bed \\
+        --OUTPUT \$bait_intervals \\
+        --SEQUENCE_DICTIONARY $dict \\
+        --TMP_DIR .
+
+    else
+        bait_intervals=$bait_intervals
+    fi
+
+    if [[ "$target_intervals" == *.bed || "$target_intervals" == *.bed.gz ]]; then
+    target_bed=$target_intervals
+    target_intervals=\${target_bed%.bed*}.interval_list
+
+    picard \\
+        -Xmx${avail_mem}M \\
+        BedToIntervalList \\
+        --INPUT \$target_bed \\
+        --OUTPUT \$target_intervals  \\
+        --SEQUENCE_DICTIONARY $dict \\
+        --TMP_DIR .
+
+    else
+        target_intervals=$target_intervals
+    fi
+
+
     picard \\
         -Xmx${avail_mem}M \\
         CollectHsMetrics \\
         $args \\
         $reference \\
-        --BAIT_INTERVALS $bait_intervals \\
-        --TARGET_INTERVALS $target_intervals \\
+        --BAIT_INTERVALS \$bait_intervals \\
+        --TARGET_INTERVALS \$target_intervals \\
         --INPUT $bam \\
         --OUTPUT ${prefix}.CollectHsMetrics.coverage_metrics
 
@@ -58,4 +94,4 @@
         picard: \$(echo \$(picard CollectHsMetrics --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d:)
     END_VERSIONS
     """
-}
+}
************************************************************
