Changes in module 'nf-core/star/align'
--- modules/nf-core/star/align/main.nf
+++ modules/nf-core/star/align/main.nf
@@ -8,12 +8,7 @@
         'biocontainers/mulled-v2-1fa26d1ce03c295fe2fdcf85831a92fbcbd7e8c2:ded3841da0194af2701c780e9b3d653a85d27489-0' }"
 
     input:
-    tuple val(meta), path(reads, stageAs: "input*/*")
-    tuple val(meta2), path(index)
-    tuple val(meta3), path(gtf)
-    val star_ignore_sjdbgtf
-    val seq_platform
-    val seq_center
+    tuple val(meta), path(reads, stageAs: "input*/*"), path(index), path(gtf)
 
     output:
     tuple val(meta), path('*Log.final.out')   , emit: log_final
@@ -42,10 +37,7 @@
     def prefix = task.ext.prefix ?: "${meta.id}"
     def reads1 = [], reads2 = []
     meta.single_end ? [reads].flatten().each{reads1 << it} : reads.eachWithIndex{ v, ix -> ( ix & 1 ? reads2 : reads1) << v }
-    def ignore_gtf      = star_ignore_sjdbgtf ? '' : "--sjdbGTFfile $gtf"
-    def seq_platform    = seq_platform ? "'PL:$seq_platform'" : ""
-    def seq_center      = seq_center ? "'CN:$seq_center'" : ""
-    def attrRG          = args.contains("--outSAMattrRGline") ? "" : "--outSAMattrRGline 'ID:$prefix' $seq_center 'SM:$prefix' $seq_platform"
+    def ignore_gtf      = $gtf ? "--sjdbGTFfile $gtf" : ""
     def out_sam_type    = (args.contains('--outSAMtype')) ? '' : '--outSAMtype BAM Unsorted'
     def mv_unsorted_bam = (args.contains('--outSAMtype BAM Unsorted SortedByCoordinate')) ? "mv ${prefix}.Aligned.out.bam ${prefix}.Aligned.unsort.out.bam" : ''
     """
@@ -56,7 +48,6 @@
         --outFileNamePrefix $prefix. \\
         $out_sam_type \\
         $ignore_gtf \\
-        $attrRG \\
         $args
 
     $mv_unsorted_bam

************************************************************
