# nf-cmgg/preprocessing: Output

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

### Sample information

A separate directory will be created in the output directory for each sample containing all output files for that sample.

<details markdown="1">
<summary>Output files</summary>

- `SAMPLE`
  - `SAMPLE.CollectHsMetrics.coverage_metrics`: The coverage metrics calculated by `picard CollectHsMetrics`
  - `SAMPLE.CollectMultipleMetrics.alignment_summary_metrics`: The alignment summary metrics calculated by `picard CollectMultipleMEtrics`
  - `SAMPLE.CollectMultipleMetrics.base_distribution_by_cycle_metrics`: The base distribution by cycle metrics calculated by `picard CollectMultipleMetrics`
  - `SAMPLE.CollectMultipleMetrics.base_distribution_by_cycle.pdf`: PDF file containing the base distribution by cycle metrics
  - `SAMPLE.CollectMultipleMetrics.quality_by_cycle_metrics`: The quality by cycle metrics calculated by `picard CollectMultipleMetrics`
  - `SAMPLE.CollectMultipleMetrics.quality_by_cycle.pdf`: PDF file containing the quality by cycle metrics
  - `SAMPLE.CollectMultipleMetrics.quality_distribution_metrics`: The quality by distribution metrics calculated by `picard CollectMultipleMetrics`
  - `SAMPLE.CollectMultipleMetrics.quality_distribution.pdf`: PDF file containing the quality distribution metrics
  - `SAMPLE.CollectMultipleMetrics.read_length_histogram.pdf`: A histogram detailing the read lenghts made with `picard CollectMultipleMetrics`
  - `SAMPLE.coverage.txt`: The coverage metrics calculated by `samtools coverage`
  - `SAMPLE.cram`: The CRAM file generated by the aligner
  - `SAMPLE.cram.crai`: The index of the CRAM file
  - `SAMPLE.cram.md5`: The md5sum of the CRAM file
  - `SAMPLE.duplicate_metrics.txt`: The duplicate metrics calculated by `samtools mardup`
  - `SAMPLE.fastp.html`: The HTML file visualising the metrics calculated by `fastp`
  - `SAMPLE.fastp.json`: The JSON file containing the metrics calculated by `fastp`
  - `SAMPLE.flagstat`: The quality control metrics calculated by `samtools flagstat`
  - `SAMPLE.idxstats`: The quality control metrics calculated by `samtools idxstats`
  - `SAMPLE.mosdepth.global.dist.txt`: The global read distribution metrics calculated by `mosdepth`
  - `SAMPLE.mosdepth.region.dist.txt`: The regional read distribution metrics calculated by `mosdepth` (only when a ROI BED file has been given)
  - `SAMPLE.mosdepth.summary.txt`: The summary of the `mosdepth` metrics
  - `SAMPLE.per-base.bed.gz`: The per-base depth calculated by `mosdepth`
  - `SAMPLE.per-base.bed.gz.csi`: The index of the per-base depth BED file
  - `SAMPLE.quantized.bed.gz`: The quantized BED file, showing how well covered each genomic region is, calculated by `mosdepth`
  - `SAMPLE.quantized.bed.gz.csi`: The index of the quantized BED file
  - `SAMPLE.regions.bed.gz`: The regional BED file, showing how well covered the requested regions are, calculated by `mosdepth` (only when a ROI BED file has been given)
  - `SAMPLE.regions.bed.gz.csi`: The index of the regional BED file
  - `SAMPLE.stats`: General statistics for the sample
  </details>

### Extra outputs for flowcell inputs

Some additional files will be created when a flowcell input has been used.

<details markdown="1">
<summary>Output files</summary>

- `InterOp/`: Directory containing `bins` for each sample
- `Reports/`: Quality control reports for the flowcell run

</details>

### MultiQC

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
