Changes in module 'nf-core/samtools/cat'
--- modules/nf-core/samtools/cat/main.nf
+++ modules/nf-core/samtools/cat/main.nf
@@ -11,9 +11,9 @@
     tuple val(meta),  path(input_files, stageAs: "?/*")
 
     output:
-    tuple val(meta), path("${prefix}.bam") , optional:true, emit: bam
-    tuple val(meta), path("${prefix}.cram"), optional:true, emit: cram
-    path  "versions.yml"                                  , emit: versions
+    tuple val(meta), path("*.bam") , optional:true, emit: bam
+    tuple val(meta), path("*.cram"), optional:true, emit: cram
+    path  "versions.yml"           , emit: versions
 
 
     when:

************************************************************
